datumaro/util/scope.py

```diff
@@ -0,0 +1,171 @@
+        """
+
+        if args or kwargs or fwd_kwargs:
+            if fwd_kwargs:
+                kwargs.update(fwd_kwargs)
+            callback = partial(callback, *args, **kwargs)
+
+        name = name or hash(callback)
+        assert name not in self._handlers, "Callback is already registered"
+
+        handler = self.Handler(callback,
+            enabled=enabled, ignore_errors=ignore_errors)
+        self._handlers[name] = handler
+        self._exit_stack.callback(handler)
+        return name
+
+    def add(self, cm: ContextManager) -> Any:
+        """
+        Enters a context manager and adds it to the exit stack.
+        """
+
+        return self._exit_stack.enter_context(cm)
+
+    def enable(self, name=None):
+        if name:
+            self._handlers[name].enabled = True
+        else:
+            self.enabled = True
+
+    def disable(self, name=None):
+        if name:
+            self._handlers[name].enabled = False
+        else:
+            self.enabled = False
+
+    def clean(self):
+        self.__exit__()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type=None, exc_value=None, exc_traceback=None):
+        if not self.enabled:
+            return
+
+        try:
+            if exc_type:
+                self._error_stack.__exit__(exc_type, exc_value, exc_traceback)
+        finally:
+            self._exit_stack.__exit__(exc_type, exc_value, exc_traceback)
+
+    @classmethod
+    def current(cls) -> 'Scope':
+        return cls._thread_locals.current
+
+    @contextmanager
+    def as_current(self):
+        previous = getattr(self._thread_locals, 'current', None)
+        self._thread_locals.current = self
+        try:
+            yield
+        finally:
+            self._thread_locals.current = previous
+
+@optional_arg_decorator
+def scoped(func, arg_name=None):
+    """
+    A function decorator, which allows to do actions with the current scope,
+    such as registering error and exit callbacks and context managers.
+    """
+
+    @wraps(func)
+    def wrapped_func(*args, **kwargs):
+        with Scope() as scope:
+            if arg_name is None:
+                with scope.as_current():
+                    ret_val = func(*args, **kwargs)
+            else:
+                kwargs[arg_name] = scope
+                ret_val = func(*args, **kwargs)
+            return ret_val
+
+    return wrapped_func
+
+# Shorthands for common cases
+def on_error_do(callback, *args, ignore_errors=False):
+    return Scope.current().on_error_do(callback, *args,
+        ignore_errors=ignore_errors)
+on_error_do.__doc__ = Scope.on_error_do.__doc__
+
+def on_exit_do(callback, *args, ignore_errors=False):
+    return Scope.current().on_exit_do(callback, *args,
+        ignore_errors=ignore_errors)
+on_exit_do.__doc__ = Scope.on_exit_do.__doc__
+
+def add(cm: ContextManager):
+    return Scope.current().add(cm)
+add.__doc__ = Scope.add.__doc__
+
+def current():
```

Comment:
I don't think `add` and `current` are good names for module-scope functions. They're too ambiguous if imported using `from ... import` (and I don't think you want to do `import datumaro.util.scope; datumaro.util.scope.add(...)`).

`current` in particular doesn't seem necessary at all. `Scope.current()` is already short enough.

Reply:
I want these two functions to be used only as `import datumaro.util.scope as scope; scope.add(); scope.current()`. Dropping all the free functions and making the `Scope` class a decorator with static public methods is an option, but it means mixing 3 interfaces in a single entity, which I don't want to do.

Reply:
`scope.current()` still seems pointless - `Scope.current()` is the same number of characters and doesn't require an extra wrapper function.

As for `add`, IMO, it would be more convenient to rename it to `scope_add` - then you could use the same style of import for it as for everything else.

Reply:
I'm not sure about `scope_add`, but I don't dislike it enough.